Common/src/geometry/CPhysicalGeometry.cpp

```diff
@@ -11041,12 +11038,15 @@ void CPhysicalGeometry::SetWallDistance(const CConfig *config, CADTElemClass *Wa
 
       WallADT->DetermineNearestElement(nodes->GetCoord(iPoint), dist, markerID, elemID, rankID);
 
-      nodes->SetWall_Distance(iPoint, min(dist,nodes->GetWall_Distance(iPoint)));
+      if(dist < nodes->GetWall_Distance(iPoint)){
+        nodes->SetWall_Distance(iPoint, dist);
 
-      if (config->GetnRoughWall() > 0) {
-        auto index = GlobalMarkerStorageDispl[rankID] + markerID;
-        auto localRoughness = GlobalRoughness_Height[index];
-        nodes->SetRoughnessHeight(iPoint, localRoughness);
+        if (config->GetnRoughWall() > 0) {
+          const CPhysicalGeometry* ph_geometry = dynamic_cast<const CPhysicalGeometry*>(geometry);
+          auto index = ph_geometry->GlobalMarkerStorageDispl[rankID] + markerID;
```

Comment:
Thanks Max, I agree with your assessment of the issue, but perhaps this could be done avoiding passing geometries into geometry just to retrieve a couple of fields (which then requires casting those geometries).

For example, CPhysicalGeometry::SetGlobalMarkerRoughness could become a free function, or a member of CConfig, that returns a custom type with interface `(rank,marker)`, objects of this type would be prepared in CGeometry::ComputeWallDistance and passed to SetWallDistance instead of entire geometries (if they are `.empty()` the roughness would not be set).

Alternatively, this would be more generic, and possibly useful for the future, we could pass `jZone` to SetWallDistance and store, in the nodes (i.e. CPoint), `zone,rank,marker` associated with the wall distance. Then, a separate method e.g. `SetWallRoughness`, would use this info to populate the wall roughness. This method would be called from CGeometry::ComputeWallDistance and it would also make use of above change to CPhysicalGeometry::SetGlobalMarkerRoughness (i.e. "freeing it" or moving to CConfig).

Minimum requirement IMO would be to add a virtual method to CGeometry to obtain the roughness from (rank,marker) instead of dynamic cast to access member variables.